### PR TITLE
feat: kubernetes mvp

### PR DESCRIPTION
**Pull request by florian-hoenicke**
MVP implementation of simple

### CODE REVIEW

**Comment by bwanglzu** [comment]
i like the usage of k8s, while i doubt if it should exist in core, it looks wired to run `f.deploy('destination')`. We can't see a clear boundary between development and deployment. In flask/django or such web apps, you don't use `app.deploy('destination')`. How do you think?

**Comment by mapleeit** [comment]
I really like this idea. It's pretty cool.

So the basic idea is two steps: 

1. Creating multiple individual pods as k8s deployment.
2. Then put them together within one flow with all pods which are just generated as external pods.

Am I right? @florian-hoenicke  It's not so cloud native, but it's quite efficient.

**Review by mapleeit — COMMENTED** [review]
I really like this idea. It's pretty cool.

So the basic idea is two steps: 

1. Creating multiple individual pods as k8s deployment.
2. Then put them together within one flow with all pods which are just generated as external pods.

Am I right? @florian-hoenicke  It's not so "cloud native", but it's quite efficient. Not sure about the performance, will gateway pod be the bottleneck? How to set the number of replicas of gateway?

**Comment by florian-hoenicke** [comment]
@mapleeit Thanks for your comment. I even think it is the native way of deployment on kubernetes. we can think of each executer as being a microservice which can have its own deplyment cycle and can be scaled individually.

**Comment by florian-hoenicke** [comment]
The gateway can be replicated as well. We could introduce a new parameter in the flow object to set the gateway replicas

**Comment by JoanFM** [comment]
To enforce the vision of Flow as a declarative object rather than an imperative one I would suggest that every Pod belongs to k8s development independently, so that u can have some Pods locally and some on K8s. We could even have it encoded in the `uses` with a prefix or something

**Comment by JoanFM** [comment]
> The gateway can be replicated as well. We could introduce a new parameter in the flow object to set the gateway replicas

if u pass any argument to the Flow it is passed to the gateway Pod, so it should already kind of work

File: jina/flow/base.py [hunk 1]
```diff
@@ -1563,3 +1568,137 @@ def update_network_interface(self, **kwargs):
...
+        """
+
+        if deployment_type == 'k8s':
+            # self.logger.info(f'✨ Deploy Flow on Kubernetes...')
+            namespace = self.args.name
+            self.logger.info(f'📦\tCreate Namespace {namespace}')
+            kubernetes_tools.create('namespace', {'name': namespace})
+
+            pod_to_args = {}
+            for pod_name, pod in self._pod_nodes.items():
```
Comment (JoanFM) [inline jina/flow/base.py:1589]: I would like to remove the logic from here and move it to a `K8sPod` object

File: jina/kubernetes/kubernetes_tools.py [hunk 2]
```diff
@@ -1,86 +1,130 @@
...
-                        ]
-                    ),
-                )
-            ]
-        ),
-    )
-    # Creation of the Deployment in specified namespace
-    # (Can replace "default" with a namespace you may have created)
-    networking_v1_beta1_api.create_namespaced_ingress(namespace=namespace, body=body)
+class ClientsSingelton:
```
Comment (JoanFM) [inline jina/kubernetes/kubernetes_tools.py:20]: typo: singleton

Comment (JoanFM) [reply jina/kubernetes/kubernetes_tools.py:20]: How do u make sure this is a `singleton`? I think the only way in Python is to implement a `metaclass` that controls the actual instantiation

File: extra-requirements.txt [hunk 3]
```diff
@@ -69,5 +69,6 @@ bs4:                        cicd
 aiostream:                  standard, daemon, devel
 jsonschema:                 cicd
 dgl>=0.6.1:                 devel
+kubernetes==18.20.0b1:      core
```
Comment (JoanFM) [inline extra-requirements.txt:72]: lets remember to remove this from core requirements. We should maybe add a new tag

File: jina/enums.py [hunk 4]
```diff
@@ -321,6 +321,12 @@ def is_keep(self) -> bool:
         """
         return self.value in {0, 2}
 
+class InfrastructureType(BetterEnum):
+    """Infrastructure where the Flow runs on."""
+
+    LOCAL = 0  # flow is started locally
```
Comment (JoanFM) [inline jina/enums.py:327]: I would change the name, I would call Jina vs K8s. Local is not correct because Jina can be distributed

Comment (JoanFM) [reply jina/enums.py:327]: Also in the comments do not refer to Flow but to Pods or services

File: jina/hubble/hubio.py [hunk 5]
```diff
@@ -541,7 +541,7 @@ def _get_prettyprint_usage(self, console, usage):
 
     @staticmethod
     @disk_cache_offline(cache_file=str(_cache_file))
-    def _fetch_meta(
+    def fetch_meta(
```
Comment (JoanFM) [inline jina/hubble/hubio.py:544]: why does this need to be public?

File: jina/kubernetes/generic_gateway/gateway.py [hunk 6]
```diff
@@ -0,0 +1,14 @@
+import base64
+import sys
+
+from jina import Flow
+
+print('args', sys.argv)
```
Comment (JoanFM) [inline jina/kubernetes/generic_gateway/gateway.py:6]: A gateway should not be handled by a dummy Flow. It should be a Kubernetes deployment where entrypoint is "jina gateway". Btw, jina gateway needs to be offered in the cli/api, will do tomorrow

File: jina/parsers/flow.py [hunk 7]
```diff
@@ -34,6 +35,21 @@ def mixin_flow_features_parser(parser):
     )
 
 
+def mixin_k8s_parser(parser):
```
Comment (JoanFM) [inline jina/parsers/flow.py:38]: this should be in Pod parsing

File: jina/types/routing/table.py [hunk 8]
```diff
@@ -152,7 +153,7 @@ def add_edge(self, from_pod: str, to_pod: str, send_as_bind: bool = False) -> No
         self._get_target_pod(from_pod).add_edge(to_pod, send_as_bind)
         self._get_target_pod(to_pod).expected_parts += 1
 
-    def add_pod(self, pod_name: str, pod: 'BasePod') -> None:
+    def add_pod(self, pod_name: str, flow_name, pod: 'BasePod') -> None:
```
Comment (JoanFM) [inline jina/types/routing/table.py:156]: why is flow name needed?

**Review by JoanFM — CHANGES_REQUESTED** [review]
Just noting things before pairing. Also kubernetes folder should go into peapods

File: jina/hubble/hubio.py (hunk 5 above)
Comment (florian-hoenicke) [reply jina/hubble/hubio.py:533]: we need to call it outside the class when getting the docker image for the hub executor

File: jina/peapods/pods/k8s.py [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+            f'"--grpc-data-requests", '
+            f'"--runtime-cls", "GRPCDataRuntime", '
+            f'{kubernetes_deployment.get_cli_params(self.args, ("pod_role",))}]',
+            logger=JinaLogger(f'deploy_{self.name}'),
+            replicas=1,
+            pull_policy='Always',
+            init_container=None,
+        )
+
+    def _deploy_runtime(self, deployment_args, replicas, k8s_namespace, deployment_id):
```
[the remark this answers is not in the document]
Comment (florian-hoenicke) [reply jina/peapods/pods/k8s.py:80]: Thanks, I created a ticket with the clean up tasks accordingly
https://github.com/jina-ai/jina/issues/3289

File: jina/peapods/pods/k8slib/kubernetes_tools.py [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+
+    """
+
+    def __init__(self):
+        self.__k8s_client = None
+        self.__v1 = None
+        self.__beta = None
+        self.__networking_v1_beta1_api = None
+
+    def __instantiate(self):
```
[the remark this answers is not in the document]
Comment (florian-hoenicke) [reply jina/peapods/pods/k8slib/kubernetes_tools.py:20]: Thanks, I added this to the cleanup tasks: 
https://github.com/jina-ai/jina/issues/3289

File: jina/peapods/pods/k8slib/template/namespace.yml [hunk 11]
```diff
@@ -0,0 +1,10 @@
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: {name}
+#  labels:
+#    istio-injection: enabled
```
[the remark this answers is not in the document]
Comment (florian-hoenicke) [reply jina/peapods/pods/k8slib/template/namespace.yml]: thanks, removed

File: jina/parsers/flow.py [hunk 12]
```diff
@@ -42,6 +43,21 @@ def mixin_flow_features_parser(parser):
...
+
+    :param parser: the parser configure
+    """
+    gp = add_arg_group(parser, title='Kubernetes Feature')
+    gp.add_argument(
+        '--infrastructure',
+        type=InfrastructureType.from_string,
+        choices=list(InfrastructureType),
+        default=InfrastructureType.JINA,
+        help='Infrastructure where the Flow runs on. Currently, `local` and `k8s` are supported',
```
[the remark this answers is not in the document]
Comment (florian-hoenicke) [reply jina/parsers/flow.py]: thanks, done

File: jina/flow/base.py [hunk 13]
```diff
@@ -837,11 +874,23 @@ def _get_routing_table(self) -> RoutingTable:
                     pod.head_args.hosts_in_connect.append(
                         graph._get_target_pod(start).full_out_address
                     )
-
-                    graph.add_edge(start, end, True)
+                    if end not in graph.pods:
+                        end = end + '_head'
+                    graph.add_edge(start, end, False)
```
Comment (maximilianwerk) [inline jina/flow/base.py:879]: Why is this `False`? I think this is the reason for `tests/distributed/test_local_flow_use_remote_executor/test_integration.py::test_local_flow_use_external_executor` not passing

Comment (florian-hoenicke) [reply jina/flow/base.py]: Thanks, I changed it back to True

File: setup.py [hunk 14]
```diff
@@ -141,6 +141,7 @@ def get_extra_requires(path, add_all=True):
 setup(
     name=pkg_name,
     packages=find_packages(),
+    package_data={'package': ['']},
```
Comment (maximilianwerk) [inline setup.py:144]: Is this change really needed?

Comment (florian-hoenicke) [reply setup.py:144]: thanks, it was not needed

**Review by maximilianwerk — DISMISSED** [review]
LGTM

**Review by jacobowitz — CHANGES_REQUESTED** [review]
Some smaller questions/suggestions. Nothing major.

File: jina/flow/base.py [hunk 15]
```diff
@@ -808,27 +821,54 @@ def _get_gateway_target(self, prefix):
             },
         )
 
+    # TODO needs to be refactored
```
[the remark this answers is not in the document]
Comment (florian-hoenicke) [reply jina/flow/base.py:824]: you are right. This was not so informative. In the next commit, I will link this ticket:
https://github.com/jina-ai/jina/issues/3280

File: jina/flow/base.py [hunk 16]
```diff
@@ -925,7 +974,8 @@ def build(self, copy_flow: bool = False) -> 'Flow':
         op_flow._set_initial_dynamic_routing_table()
 
         for pod in op_flow._pod_nodes.values():
-            pod.args.host = self._resolve_host(pod.args.host)
+            pass
+        pod.args.host = self._resolve_host(pod.args.host)
```
[the remark this answers is not in the document]
Comment (florian-hoenicke) [reply jina/flow/base.py:978]: oh, I forgot to change it back. Thanks

File: jina/flow/base.py [hunk 17]
```diff
@@ -1344,45 +1393,48 @@ def __iter__(self):
...
-                f'{self.address_private}:{self.port_expose}',
-                'cyan',
-                attrs='underline',
-            ),
-        ]
-        if self.address_public:
-            address_table.append(
-                f'\t🌐 Public address:\t'
+        if self.args.infrastructure == InfrastructureType.K8S:
+            self.logger.info('🎉 Kubernetes deployment done!')
```
[the remark this answers is not in the document]
Comment (florian-hoenicke) [reply jina/flow/base.py:1395]: Thanks, I created a related ticket https://github.com/jina-ai/jina/issues/3298

File: jina/peapods/pods/k8s.py [hunk 18]
```diff
@@ -0,0 +1,285 @@
...
+        init_container_args = kubernetes_deployment.get_init_container_args(self)
+        uses_metas = kubernetes_deployment.dictionary_to_cli_param(
+            {'pea_id': deployment_id}
+        )
+        uses_with = kubernetes_deployment.dictionary_to_cli_param(
+            deployment_args.uses_with
+        )
+        uses_with_string = f'"--uses-with", "{uses_with}", ' if uses_with else ''
+        if image_name == __default_executor__:
+            image_name = 'gcr.io/jina-showcase/custom-jina:latest'
```
[the remark this answers is not in the document]
Comment (florian-hoenicke) [reply jina/peapods/pods/k8s.py]: Thanks, replaced it with the current jina version in case it exists already on docker hub - otherwise it uses master.

File: jina/peapods/pods/k8s.py [hunk 19]
```diff
@@ -0,0 +1,285 @@
...
+    def head_zmq_identity(self) -> bytes:
+        """zmq identity is not needed for k8s deployment
+
+        :return: zmq identity
+        """
+        return b''
+
+    @property
+    def deployments(self) -> List[Dict]:
+        """Deployment information for the routing table creation.
```
[the remark this answers is not in the document]
Comment (florian-hoenicke) [reply jina/peapods/pods/k8s.py:235]: makes sense - changed

File: jina/peapods/pods/k8s.py [hunk 20]
```diff
@@ -0,0 +1,285 @@
...
+
+        :return: list of dictionaries defining the attributes used by the routing table
+        """
+        res = []
+        if self.args.name == 'gateway':
+            name = kubernetes_deployment.to_dns_name(self.name)
+            res.append(
+                {
+                    'name': f'{self.name}',
+                    'head_host': f'{name}.{self.args.k8s_namespace}.svc.cluster.local',
```
[the remark this answers is not in the document]
Comment (florian-hoenicke) [reply jina/peapods/pods/k8s.py:245]: You mean making `svc.cluster.local` configurable?
I would wait for people asking for this to see if this is actually required in some cases.

File: jina/resources/k8s/template/deployment-init.yml [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+        command: {command}
+        args: {args}
+        ports:
+          - containerPort: {port_expose}
+          - containerPort: {port_in}
+          - containerPort: {port_out}
+          - containerPort: {port_ctrl}
+        env:
+        - name: JINA_LOG_LEVEL
+          value: debug
```
[the remark this answers is not in the document]
Comment (florian-hoenicke) [reply jina/resources/k8s/template/deployment-init.yml:42]: Thanks, I created a ticket for this:
https://github.com/jina-ai/jina/issues/3299

File: jina/peapods/pods/k8s.py [hunk 22]
```diff
@@ -0,0 +1,285 @@
...
+                )
+        return self
+
+    def wait_start_success(self):
+        """Not implemented. It should wait until the deployment is up and running"""
+        pass
+
+    def close(self):
+        """Not implemented. It should delete the namespace of the flow"""
+        kubernetes_tools.delete_namespace(self.args.k8s_namespace)
```
[the remark this answers is not in the document]
Comment (florian-hoenicke) [reply jina/peapods/pods/k8s.py:180]: Thanks, I removed it. The implementation did not work.

File: jina/flow/base.py [hunk 23]
```diff
@@ -842,6 +882,15 @@ def _get_routing_table(self) -> RoutingTable:
                 else:
                     graph.add_edge(start, end)
 
+        # ADD INTERNAL POD BETWEEN EDGES
```
[the remark this answers is not in the document]
Comment (florian-hoenicke) [reply jina/flow/base.py]: Good point. I changed the comment

**Review by jacobowitz — APPROVED** [review]
looks good! lets merge and try it out :)

**Review by hanxiao — APPROVED** [review]
